pytestqt/modeltest.py

```diff
@@ -0,0 +1,607 @@
+        next_data = self._model.data(self._model.index(end + 1, 0, parent))
+        c = _Changing(parent=parent, oldSize=self._model.rowCount(parent),
+                      last=last_data, next=next_data)
+        self._remove.append(c)
+
+    def _on_rows_removed(self, parent, start, end):
+        """Confirm that what was said was going to happen actually did."""
+        c = self._remove.pop()
+        last_data = self._model.data(self._model.index(start - 1, 0, c.parent))
+        next_data = self._model.data(self._model.index(start, 0, c.parent))
+
+        assert c.parent == parent
+        assert c.oldSize - (end - start + 1) == self._model.rowCount(parent)
+        assert c.last == last_data
+        assert c.next == next_data
+
+    def _on_data_changed(self, topLeft, bottomRight):
+        assert topLeft.isValid()
+        assert bottomRight.isValid()
+        commonParent = bottomRight.parent()
+        assert topLeft.parent() == commonParent
+        assert topLeft.row() <= bottomRight.row()
+        assert topLeft.column() <= bottomRight.column()
+        rowCount = self._model.rowCount(commonParent)
+        columnCount = self._column_count(commonParent)
+        assert bottomRight.row() < rowCount
+        assert bottomRight.column() < columnCount
+
+    def _on_header_data_changed(self, orientation, start, end):
+        assert orientation in [QtCore.Qt.Horizontal, QtCore.Qt.Vertical]
+        assert start >= 0
+        assert end >= 0
+        assert start <= end
+        if orientation == QtCore.Qt.Vertical:
+            itemCount = self._model.rowCount()
+        else:
+            itemCount = self._column_count()
+        assert start < itemCount
+        assert end < itemCount
+
+    def _column_count(self, parent=QtCore.QModelIndex()):
+        """
+        Workaround for the fact that ``columnCount`` is a private method in
+        QAbstractListModel/QAbstractTableModel subclasses and PySide does not
+        provide any way to work around that limitation. PyQt lets us "cast"
+        back to the super class to access these private methods.
+        """
+        if isinstance(self._orig_model, QAbstractListModel) and USING_PYSIDE:
+            return 1 if parent == QtCore.QModelIndex() else 0
+        else:
+            return self._model.columnCount(parent)
+
+    def _parent(self, index):
+        """
+        .. see:: ``_column_count``
+        """
+        model_types = (QAbstractListModel, QAbstractTableModel)
+        if isinstance(self._orig_model, model_types) and USING_PYSIDE:
+            return QtCore.QModelIndex()
+        else:
+            return self._model.parent(index)
+
+    def _has_children(self, parent=QtCore.QModelIndex()):
+        """
+        .. see:: ``_column_count``
+        """
+        model_types = (QAbstractListModel, QAbstractTableModel)
+        if isinstance(self._orig_model, model_types) and USING_PYSIDE:
+            return parent == QtCore.QModelIndex() and self._model.rowCount() > 0
```

Comment:
This looks wrong - it would mean any item of a `QAbstractListModel` and `QAbstractTableModel` would have children, right?

This should probably be something like the C++ source:

```cpp
bool QAbstractTableModel::hasChildren(const QModelIndex &parent) const
{
    if (parent.model() == this || !parent.isValid())
        return rowCount(parent) > 0 && columnCount(parent) > 0;
    return false;
}

bool QAbstractListModel::hasChildren(const QModelIndex &parent) const
{
    return parent.isValid() ? false : (rowCount() > 0);
}
```

I'm not sure why `QAbstractListModel::hasChildren` doesn't check the model of the index - but I think doing this is undefined behaviour anyways.